Source/JavaScriptCore/jit/AssemblyHelpers.cpp

```diff
@@ -409,8 +409,13 @@ void AssemblyHelpers::emitNonNullDecodeStructureID(RegisterID source, RegisterID
     lshift64(source, TrustedImm32(StructureID::encodeShiftAmount), dest);
 #elif CPU(ADDRESS64)
     // This could use BFI on arm64 but that only helps if the start of structure heap is encodable as a mov and not as an immediate in the add so it's probably not super important.
-    and32(TrustedImm32(StructureID::structureIDMask), source, dest);
-    add64(TrustedImm64(g_jscConfig.startOfStructureHeap), dest);
+    if constexpr (structureHeapAddressSize >= 4 * GB) {
+        move(source, dest);
```

Comment:
source is guaranteed to be 32-bit zero extended coming in? Should this just be move32?

Maybe we should assert heapSize <= 4gb too. so this can just be == 4gb

Reply:
At least, in the current code, yes, it is ensured. I'll change the function name to `emitNonNullDecodeZeroExtendedStructureID` to ensure that incoming structure ID needs to be zero-extended.

Reply:
I'll add assertion about `<= 4GB` check.